scipy/signal/_sosfilt.pyx

```diff
@@ -21,38 +24,63 @@ ctypedef fused DTYPE_t:
 @cython.cdivision(True)
 @cython.boundscheck(False)
 @cython.wraparound(False)
-def _sosfilt(DTYPE_t [:, ::1] sos,
-             DTYPE_t [:, ::1] x,
-             DTYPE_t [:, :, ::1] zi):
+cdef void _sosfilt_float(DTYPE_floating_t [:, ::1] sos,
+                         DTYPE_floating_t [:, ::1] x,
+                         DTYPE_floating_t [:, :, ::1] zi) nogil:
     # Modifies x and zi in place
     cdef Py_ssize_t n_signals = x.shape[0]
     cdef Py_ssize_t n_samples = x.shape[1]
     cdef Py_ssize_t n_sections = sos.shape[0]
     cdef Py_ssize_t i, n, s
-    cdef DTYPE_t x_n
-    cdef DTYPE_t [:, ::1] b = sos[:, :3]
-    # We ignore sos[:, 3] here because _validate_sos guarantees it's 1.
-    cdef DTYPE_t [:, ::1] a = sos[:, 4:]
+    cdef DTYPE_floating_t x_new, x_cur
+    cdef DTYPE_floating_t[:, ::1] zi_slice
+
+    # jumping through a few memoryview hoops to reduce array lookups,
+    # the original version is still in the gil version below.
+    for i in xrange(n_signals):
+        zi_slice = zi[i, :, :]
+        for n in xrange(n_samples):
+
+            x_cur = x[i, n]
+
+            for s in xrange(n_sections):
+                x_new = sos[s, 0] * x_cur + zi_slice[s, 0]
+                zi_slice[s, 0] = (sos[s, 1] * x_cur - sos[s, 4] * x_new
+                                  + zi_slice[s, 1])
+                zi_slice[s, 1] = sos[s, 2] * x_cur - sos[s, 5] * x_new
+                x_cur = x_new
+
+            x[i, n] = x_cur
+
+
+@cython.cdivision(True)
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def _sosfilt_object(object [:, ::1] sos,
+                    object [:, ::1] x,
+                    object [:, :, ::1] zi):
+    # Modifies x and zi in place
+    cdef Py_ssize_t n_signals = x.shape[0]
+    cdef Py_ssize_t n_samples = x.shape[1]
+    cdef Py_ssize_t n_sections = sos.shape[0]
+    cdef Py_ssize_t i, n, s
+    cdef object x_n
+    for i in xrange(n_signals):
+        for n in xrange(n_samples):
+            for s in xrange(n_sections):
+                x_n = x[i, n]  # make a temporary copy
+                # Use direct II transposed structure:
+                x[i, n] = sos[s, 0] * x_n + zi[i, s, 0]
+                zi[i, s, 0] = (sos[s, 1] * x_n - sos[s, 4] * x[i, n] +
+                               zi[i, s, 1])
+                zi[i, s, 1] = (sos[s, 2] * x_n - sos[s, 5] * x[i, n])
+
+
+cpdef void _sosfilt(DTYPE_t [:, ::1] sos,
+             DTYPE_t [:, ::1] x,
+             DTYPE_t [:, :, ::1] zi) nogil:
```

Comment:
This doesn't release the GIL any more. Quoting the [docs](https://cython.readthedocs.io/en/latest/src/userguide/external_C_code.html#declaring-a-function-as-callable-without-the-gil)
>The `nogil` function annotation declares that it is safe to call the function without the GIL. It is perfectly allowed to execute it while holding the GIL. The function does not in itself release the GIL if it is held by the caller.

I'll send a follow-up PR.

Reply:
Thanks @peterbell10 , I missed that!